Fix Rails 6.1 database config

joshleblanc opened the pull request:
This PR expands on #854.

In addition to #854, it fixes a problem with `rake db:create` failing to run, as well as fixes queries on arrays throwing.

As a happy accident, this also fixes #841

runephilosof-abtion commented:
Still does not seem to work 

This error is from the https://github.com/rails-sqlserver/ss_app, which I upgraded to Rails 6.1 https://github.com/runephilosof/ss_app/tree/rails6.1.

```
[vagrant@localhost ss_app]$ rails db:create
TinyTds::Error: Incorrect syntax near 'DATABASE'.
Couldn't create 'ss_app_development' database. Please check your configuration.
rails aborted!
ActiveRecord::StatementInvalid: TinyTds::Error: Incorrect syntax near 'DATABASE'.
/home/vagrant/ss_app/bin/rails:5:in `<top (required)>'
/home/vagrant/ss_app/bin/spring:10:in `block in <top (required)>'
/home/vagrant/ss_app/bin/spring:7:in `tap'
/home/vagrant/ss_app/bin/spring:7:in `<top (required)>'

Caused by:
TinyTds::Error: Incorrect syntax near 'DATABASE'.
/home/vagrant/ss_app/bin/rails:5:in `<top (required)>'
/home/vagrant/ss_app/bin/spring:10:in `block in <top (required)>'
/home/vagrant/ss_app/bin/spring:7:in `tap'
/home/vagrant/ss_app/bin/spring:7:in `<top (required)>'
Tasks: TOP => db:create
(See full trace by running task with --trace)
```

It also fails when running `rails db:drop`
```
[vagrant@localhost ss_app]$ rails db:drop
TinyTds::Error: Incorrect syntax near the keyword 'SET'.
Couldn't drop database 'ss_app_development'
rails aborted!
ActiveRecord::StatementInvalid: TinyTds::Error: Incorrect syntax near the keyword 'SET'.
/home/vagrant/ss_app/bin/rails:5:in `<top (required)>'
/home/vagrant/ss_app/bin/spring:10:in `block in <top (required)>'
/home/vagrant/ss_app/bin/spring:7:in `tap'
/home/vagrant/ss_app/bin/spring:7:in `<top (required)>'

Caused by:
TinyTds::Error: Incorrect syntax near the keyword 'SET'.
/home/vagrant/ss_app/bin/rails:5:in `<top (required)>'
/home/vagrant/ss_app/bin/spring:10:in `block in <top (required)>'
/home/vagrant/ss_app/bin/spring:7:in `tap'
/home/vagrant/ss_app/bin/spring:7:in `<top (required)>'
Tasks: TOP => db:drop:_unsafe
(See full trace by running task with --trace)
```

joshleblanc commented:
@runephilosof-abtion I believe you need to point your gemfile at the rails-6-1-support branch, it defaults to master.

Try something like `gem 'activerecord-sqlserver-adapter', github: "joshleblanc/activerecord-sqlserver-adapter", branch: "rails-6-1-support"`

runephilosof-abtion reviewed (approved):
@joshleblanc You are correct. How embarrassing. It is working perfectly.

runephilosof-abtion commented:
@wpolicarpo It would be nice to get this merged.
The tests are failing because of a pull rate limit with docker.com.
Rerun the tests?

Might want to authenticate to docker.com to increase the limit. But it should not delay this important PR.

vkalach commented:
@wpolicarpo any estimates when it will be merged?

wpolicarpo commented:
@joshleblanc can please you rebase your branch against `main`?

wpolicarpo commented on lib/active_record/connection_adapters/sqlserver/schema_statements.rb:
```diff
@@ -422,10 +422,12 @@ def column_definitions(table_name)
               else
                 type = case ci[:type]
                        when /smallint|int|bigint/ then ci[:_type]
+                       when /bit/ then 'smallint'
```
Not sure if that was the best approach to solve that issue. The fact that you had to change a spec rings a bell that something might be wrong there.

TinyTDS casts the `bit` data type to boolean, can't see why we should change it at the adapter level. That could lead to unexpected bugs/changes to already existing apps assuming `boolean` as `true/false` instead of `1/0`.

In any case #875 addressed the original issue already.

joshleblanc replied:
This change came from #854

wpolicarpo replied:
Right. #854 was closed via #875 and should not be introduced again.

wpolicarpo commented on lib/active_record/tasks/sqlserver_database_tasks.rb:
```diff
@@ -14,7 +14,7 @@ class SQLServerDatabaseTasks
 
       def initialize(configuration)
-        @configuration = configuration
+        @configuration = ActiveSupport::HashWithIndifferentAccess.new(configuration)
```
Why did you have to change it to a `HashWithIndifferentAccess`? Are you trying to mimic the new `using_database_configurations?` we should implement?

joshleblanc replied:
After the 6.1 upgrade, none of the string keys were resolving. All of the keys were symbols. 

Changing it to a hash with indifferent access let me fix it with the least amount of disruptions.

wpolicarpo commented on lib/active_record/connection_adapters/sqlserver/core_ext/array_handler.rb:
```diff
@@ -0,0 +1,49 @@
+            array_predicates = ranges.map { |range| predicate_builder.build(attribute, range) }
+            array_predicates.unshift(values_predicate)
+            array_predicates.inject(&:or)
+          end
```
Can you explain the reasoning behind these changes? If possible, can you open a separate PR for each issue you are trying to address? It's easier for us to review and keep track of changes in future bug reports.

joshleblanc replied:
This reverts an optimization made in rails 6.1, back to the functionality of 6.0. With the optimization present, any operations using an array would fail. 

For example, `where(id: [1,2,3])` would throw. 

I believe this was the commit that introduced the change: https://github.com/rails/rails/commit/72fd0bae5948c1169411941aeea6fef4c58f34a9

Sadly I don't remember the exact chain of events that lead me to this solution, it's been awhile.

joshleblanc commented:
Removed the 2 already-implemented fixes, updated the database tasks to use the configuration hash when required, rather than hash with indifferent access.

wpolicarpo commented on test/cases/column_test_sqlserver.rb:
```diff
@@ -99,7 +99,7 @@ def assert_obj_set_and_save(attribute, value)
       _(col.type).must_equal               :boolean
       _(col.null).must_equal               true
-      _(col.default).must_equal            true
+      _(col.default).must_equal            1
```
Need to revert this one too.
